lib/friends/introvert.rb

```diff
@@ -110,6 +110,9 @@ def add_activity(serialization:)
         @activities.unshift(activity)
 
         @output << "Activity added: \"#{activity}\""
+        if activity_sets_default_location?(activity)
```

Comment:
I'm not sure I understand the rationale for the `previous_activities.none?(&:default_location)` check in this method. Why can't this simply be:

```ruby
if activity.default_location
  @output << "Default location set to: \"#{activity.default_location}\""
end
```

?

Reply:
True, I originally thought the same thing. But  after some thinking I came across the scenario covered by [this test](https://github.com/shen-sat/friends/blob/c108c5de5fb9c7908bc73537d59a47c56d56ea22/test/commands/add/activity_spec.rb#L81).

The first time I set a default location (eg with the line `Went to _Paris_ for a holiday`), then I would want to have the line `Default location set to Paris` printed to the console. But if I add a subsequent activity like `Moved to _Paris_` (which also contains the trigger for setting a default location to `Paris`), then printing `Default location set to Paris` is unneccessary as we've already notified the user with the same message - and this is what my implementation prevents. What do you think? Happy to go with your recommendation on this!

Reply:
Hmm, reading my own comment has just raised an issue I didn't think of before. If someone adds activities in this order:

1. `Moved to Paris`
2. `Moved to Berlin`
3. `Went to Paris`

Then when I add the third activity, with my implementation, `friends` **won't** print `Default location set to Paris` :(

In which case your implemention is correct, as long as we are both happy with `Default location set to _LOCATION_` being printed to console twice if someone adds activities in the follwing order:

1. `Moved to Paris`
2. `Went to Paris`

What do you think?

Reply:
Oh! I see what you're saying—nice attention to detail.

Yes, the `to Paris`, `to Berlin`, `to Paris` example was what I was thinking of when I originally commented. Would you mind adding that as a test to make sure we _do_ print the message in that case?

I hadn't thought of the case of `to Paris`, `to Paris`, but I think you're right that it's weird to print the normal message. I actually worry about _not_ printing something though, because if I forget that the default location was set to that location already, I'd expect output and get nothing and would think something was wrong. I think the best solution might be something like this:

```ruby
if activity.default_location
  # This condition isn't quite right because it doesn't handle the case where activity is
  # inserted in the middle of our history... see below
  if @activities.drop(1).find(&:default_location).default_location == activity.default_location
    @output << "Default location already set to: \"#{activity.default_location}\""
  else
    @output << "Default location set to: \"#{activity.default_location}\""
  end
end
```

But now this is making me realize we should probably add tests for when the added activity is older and there's a newer default location. In that case I think it might be most helpful to print something like:

> Default location from **2016-01-01** to **2016-03-12** set to: "Paris"

Which is making me think we should maybe have a few different cases, something like:

```ruby
# this is all pseudocode
# note that some of these lookups may be easier to do in the `clean` method; I'm not sure.
if activity.default_location
  str = "Default location"

  # If we find one, keep searching back to make sure there isn't an earlier one.
  earlier_activity = activity before `activity` that has same default location || activity

  if activity is NOT latest activity
    str += " from #{Paint[earlier_activity.date, :bold]}"

    if later_activity changes the default location # If later_activity sets it to the same, keep searching for another later_activity that changes it
      str += " to #{Paint[later_activity, :bold]}"
    end
  end

  str += " already" if earlier_activity != activity

  @output << "#{str} set to: \"#{activity.default_location}\""
end
```

Thoughts?

Reply:
This is really interesting - I hadn't thought about the fact that a user could put in an _older_ activity! 😲

So from our discussion, there are 4 scenarios to cover:

1. User adds activity `Moved to _Paris_` with an empty activities log
2. User adds activity `Moved to _Paris_`, with `Went to _Paris_` as the last activity in the activities log
3. User adds activity `Moved to _Paris_`, with `Went to _Berlin_` as the last activity and `Went to _Paris_` as the activity before that in the activities log
4. User adds activity `Moved to _Paris_`, with a date that precedes the date of `Went to _Paris_` in the activities log

I think the pseudo code you wrote (thanks for writing it, it helped me understand the scenarios 😄 ) covers everything but scenario 3, but I can figure that out.

If I test those scenarios, do you think we'll have everything covered?

Reply:
Thanks for writing these out! Good call on case 3 not being covered. Maybe a clearer version would be:

```
# this is still psuedocode!

# We could think about moving this stable_sort to the read stage
# so we don't also have to do it in `clean`. Or we could skip it since
# @activities will only be unsorted if the user has been editing the
# file manually.
@activities = stable_sort(@activities)

if activity.default_location
  str = "Default location"

  # We use <= instead of < here because a newly-added activity is guaranteed to be
  # the latest of all the activities on that date.
  earlier_activities, later_activities = @activities.partition { |a| a.date <= activity.date }

  earlier_activity = activity
  earlier_activities.reverse_each do |a|
    next unless a.default_location

    if a.default_location == activity.default_location
      earlier_activity = a
    else
      break
    end
  end

  # We use < instead of <= here because a newly-added activity is guaranteed to be
  # the latest of all the activities on that date.
  if !later_activities.empty? && activity.date < later_activities.last.date
    str += " from #{Paint[earlier_activity.date, :bold]}"

    later_activity = later_activities.find { |a| a.default_location && a.default_location != activity.default_location }

    # I added the " to present" case just now as I think it makes some of the output clearer.
    str += " to #{Paint[later_activity&.date || 'present', :bold]}"
  end

  str += " already" if earlier_activity != activity

  @output << "#{str} set to: \"#{activity.default_location}\""
end

@activities.unshift(activity) # Add this after we've done our lookups above.
```

So in terms of scenarios to test, I find it easiest to think about in terms of the output branches. We want tests that can produce output for all of the following formats:

> Default location set to: "Paris"
> Default location already set to: "Paris"
> Default location from **2016-01-01** to **present** set to: "Paris"
> Default location from **2016-01-01** to **present** already set to: "Paris"
> Default location from **2016-01-01** to **2017-01-01** set to: "Paris"
> Default location from **2016-01-01** to **2017-01-01** already set to: "Paris"

Thoughts?

Reply:
I haven't looked through the pseudo-code properly yet (I will do soon!), but I agree with thinking about producing all the outputs you listed. The only thing to add is that the first item you listed (`Default location set to: "Paris"`) will cover scenarios 1 and 3 from my list - I can make this explicit in the tests 👍